presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java

```diff
@@ -334,6 +339,82 @@ private ConnectorTableMetadata getTableMetadata(SchemaTableName tableName)
         return columns.build();
     }
 
+    @Override
+    public TableStatistics getTableStatistics(ConnectorSession session, ConnectorTableHandle tableHandle, Constraint<ColumnHandle> constraint)
+    {
+        List<ConnectorTableLayoutResult> tableLayouts = getTableLayouts(session, tableHandle, constraint, Optional.empty());
+        if (tableLayouts.isEmpty()) {
+            // todo - return statistics matching empty table instead
+            return EMPTY_STATISTICS;
+        }
+        HiveTableLayoutHandle layout = checkType(tableLayouts.get(0), HiveTableLayoutHandle.class, "layoutHandle");
+        List<HivePartition> hivePartitions = layout.getPartitions().orElse(ImmutableList.of());
+        List<Long> knownPartitionRowCounts = hivePartitions.stream()
+                .map(this::getPartitionStatistics)
+                .map(PartitionStatistics::getNumRows)
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .collect(toList());
+
+        long knownPartitionRowCountsSum = knownPartitionRowCounts.stream().mapToLong(a -> a).sum();
+        long partitionsWithStatsCount = knownPartitionRowCounts.size();
+        long allPartitionsCount = hivePartitions.size();
+
+        TableStatistics tableStatistics = EMPTY_STATISTICS;
+        if (partitionsWithStatsCount > 0) {
+            tableStatistics = TableStatistics.builder()
+                    .setRowCount(new Estimate(1.0 * knownPartitionRowCountsSum / partitionsWithStatsCount * allPartitionsCount))
+                    .build();
+        }
+        return tableStatistics;
+    }
+
+    private PartitionStatistics getPartitionStatistics(HivePartition hivePartition)
+    {
+        String databaseName = hivePartition.getTableName().getSchemaName();
+        String tableName = hivePartition.getTableName().getTableName();
+        String partitionId = hivePartition.getPartitionId();
+        if (partitionId.equals(HivePartition.UNPARTITIONED_ID)) {
+            Optional<Table> tableOptional = metastore.getTable(databaseName, tableName);
+            Table table = tableOptional
+                    .orElseThrow(() -> new IllegalArgumentException(format("Could not get metadata for table %s.%s", databaseName, tableName)));
+            return readStatisticsFromParameters(table.getParameters());
+        }
+        else {
+            Optional<Partition> partitionOptional = metastore.getPartition(databaseName, tableName, toPartitionValues(partitionId));
```

Comment:
> This is going to be super expensive for tables that process large numbers of partitions. We're going to either:
> 
> 1. need a way to disable this

I will add a Hive specific config and session property to disable stats. Can come up useful until feature is stabilised.

> 2. encapsulate it under an API that could be batch requests to the metastore.

Batch requests are already done for more detailed column stats (ndv, min, max, number of nulls etc).
See series of commits with name starting with "Expose column statistics ..."

I will see if we can do better for basic stats which contain rowNumber. Maybe there is batch api for that and I just missed it.